crates/services/txpool_v2/src/storage/graph.rs

```diff
@@ -0,0 +1,693 @@
+use std::{
+    collections::{
+        HashMap,
+        HashSet,
+        VecDeque,
+    },
+    time::Instant,
+};
+
+use fuel_core_types::{
+    fuel_tx::{
+        input::{
+            coin::{
+                CoinPredicate,
+                CoinSigned,
+            },
+            contract::Contract,
+            message::{
+                MessageCoinPredicate,
+                MessageCoinSigned,
+                MessageDataPredicate,
+                MessageDataSigned,
+            },
+        },
+        ContractId,
+        Input,
+        Output,
+        TxId,
+        UtxoId,
+    },
+    services::txpool::PoolTransaction,
+};
+use petgraph::{
+    graph::NodeIndex,
+    prelude::StableDiGraph,
+};
+
+use crate::{
+    error::Error,
+    ports::TxPoolPersistentStorage,
+    selection_algorithms::ratio_tip_gas::RatioTipGasSelectionAlgorithmStorage,
+    storage::checked_collision::CheckedTransaction,
+};
+
+use super::{
+    RemovedTransactions,
+    Storage,
+    StorageData,
+};
+
+pub struct GraphStorage {
+    /// The configuration of the graph
+    config: GraphConfig,
+    /// The graph of transactions
+    graph: StableDiGraph<StorageData, ()>,
+    /// Coins -> Transaction that currently create the UTXO
+    coins_creators: HashMap<UtxoId, NodeIndex>,
+    /// Contract -> Transaction that currently create the contract
+    contracts_creators: HashMap<ContractId, NodeIndex>,
+}
+
+pub struct GraphConfig {
+    /// The maximum number of transactions per dependency chain
+    pub max_txs_chain_count: usize,
+}
+
+impl GraphStorage {
+    /// Create a new graph storage
+    pub fn new(config: GraphConfig) -> Self {
+        Self {
+            config,
+            graph: StableDiGraph::new(),
+            coins_creators: HashMap::new(),
+            contracts_creators: HashMap::new(),
+        }
+    }
+
+    #[cfg(test)]
+    pub fn is_empty(&self) -> bool {
+        self.graph.node_count() == 0
+            && self.coins_creators.is_empty()
+            && self.contracts_creators.is_empty()
+    }
+}
+
+impl GraphStorage {
+    fn reduce_dependencies_cumulative_gas_tip_and_chain_count(
+        &mut self,
+        root_id: NodeIndex,
+        removed_node: &StorageData,
+    ) {
+        let Some(root) = self.graph.node_weight_mut(root_id) else {
+            debug_assert!(false, "Node with id {:?} not found", root_id);
+            return;
+        };
```

Comment:
Are we okay with silently returning this in production? It feels like we should at least log an error in either scenario.

Reply:
Yeah, it would be nice to have a log=)

Reply:
Everywhere else there is a warn this one has been forgotten. Added.

Reply:
Nice, thanks!